jetstream/sponsored-suggestion-placement-phase-2.toml

```diff
@@ -0,0 +1,694 @@
+          * EXCEPT (is_ad_component),
+          component IN ('ad_carousel', 'ad_image_row', 'ad_link', 'ad_sidebar', 'ad_sitelink') AS ad_component
+        FROM
+          `mozdata.firefox_desktop.serp_events`
+        WHERE
+          submission_date BETWEEN '2024-01-16' AND '2024-04-22'
+          AND mozfun.map.get_key(experiments, 'sponsored-suggestion-placement-phase-2') IS NOT NULL
+      )
+      WHERE
+        search_engine = 'google'
+        AND is_tagged
+      GROUP BY
+        1, 2, 3, 4, 5, 6, 7
+    ),
+    combined AS (
+      -- combine urlbar and serp events into single table
+      -- keep only relevant fields from each
+      SELECT
+        glean_client_id,
+        seq,
+        event_timestamp,
+        legacy_telemetry_client_id,
+        submission_date,
+        experiments,
+        FALSE AS is_serp,
+        STRUCT(
+          exposed,
+          event_action,
+          product_selected_result,
+          (
+            SELECT
+              COUNTIF(product_result_type = 'admarketplace_sponsored') > 0
+            FROM
+              UNNEST(results)
+          ) AS has_sponsored_impression
+        ) AS urlbar_fields,
+        NULL AS serp_fields,
+      FROM
+        urlbar_with_exposed
+      UNION ALL (
+        SELECT
+          glean_client_id,
+          seq,
+          event_timestamp,
+          legacy_telemetry_client_id,
+          submission_date,
+          experiments,
+          TRUE AS is_serp,
+          NULL AS urlbar_fields,
+          STRUCT(
+            sap_source,
+            has_ads_loaded,
+            has_ad_click,
+            num_ad_clicks
+          ) AS serp_fields
+        FROM
+          serp
+      )
+    ),
+    sessions AS (
+      -- group events into 'search sessions'
+      -- sessions start with either a urlbar event or a SAP-sourced serp load
+      -- and include subsequent follow-on-type serp loads
+      SELECT
+        *,
+        COUNTIF(new_session_start) OVER (
+          PARTITION BY glean_client_id ORDER BY seq, event_timestamp
+        ) AS client_session_id,
+      FROM (
+        SELECT
+          *,
+          NOT is_serp OR serp_fields.sap_source IN ('searchbar', 'contextmenu', 'webextension', 'system') AS new_session_start
+        FROM
+          combined
+      )
+    ),
+    urlbar_sessions AS (
+      -- restrict to sessions starting from the urlbar
+      SELECT
+        *,
+        -- add session-level exposure indicator
+        FIRST_VALUE(urlbar_fields.exposed IGNORE NULLS) OVER (
+          PARTITION BY glean_client_id, client_session_id ORDER BY seq, event_timestamp
+        ) AS exposed_session
+      FROM
+        sessions
+      WHERE
+        -- drop partial sessions at the start of a window
+        client_session_id > 0
+      QUALIFY
+        COUNTIF(new_session_start AND NOT is_serp) OVER (
+          PARTITION BY glean_client_id, client_session_id
+        ) > 0
+    )
+    SELECT
+      *
+    FROM
+      urlbar_sessions
+)"""
```

Comment:
Yikes, what a query. Definitely in the realm of "this should be implemented as a table instead of calculated on the fly by Jetstream". I'm concerned about the complexity and the runtime of this query. I'm not going to block on changing that here, we can try to run as is, but something to consider in the future.

Reply:
Totally agree, and I have a task to work on this as we start to use this approach more. This is based on a prototype query I used for the sponsored suggestion phase 1 experiment (also in the experiment config). FWIW it gave the desired results, but I'm not sure about performance implications. I'm working on a proposal so we can align on the methodology, and then we can implement in a table.